core/src/test/scala/integration/kafka/api/EndToEndClusterIdTest.scala

```diff
@@ -0,0 +1,204 @@
+  * the License.  You may obtain a copy of the License at
+  *
+  * http://www.apache.org/licenses/LICENSE-2.0
+  *
+  * Unless required by applicable law or agreed to in writing, software
+  * distributed under the License is distributed on an "AS IS" BASIS,
+  * WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+  * See the License for the specific language governing permissions and
+  * limitations under the License.
+  */
+
+package kafka.api
+
+import java.util.concurrent.atomic.AtomicReference
+import java.util.{ArrayList, Properties}
+
+import kafka.common.TopicAndPartition
+import kafka.integration.KafkaServerTestHarness
+import kafka.server._
+import kafka.utils._
+import org.apache.kafka.clients.consumer._
+import org.apache.kafka.clients.producer.{KafkaProducer, ProducerConfig, ProducerRecord}
+import org.apache.kafka.common.{ClusterResource, ClusterResourceListener, TopicPartition}
+import org.apache.kafka.test.{TestUtils => _, _}
+import org.junit.Assert._
+import org.junit.{Before, Test}
+
+import scala.collection.JavaConverters._
+
+/** The test cases here verify the following conditions.
+  * 1. The ProducerInterceptor receives the cluster id after the onSend() method is called.
+  * 2. The Serializer receives the cluster id before the serialize() method is called.
+  * 3. The producer MetricReporter receives the cluster id  after send() method is called on KafkaProducer.
+  * 4. The ConsumerInterceptor receives the cluster id before the onConsume() method.
+  * 5. The Deserializer receives the cluster id before the deserialize() method is called.
+  * 6. The consumer MetricReporter receives the cluster id  after poll() is called on KafkaConsumer.
+  * 7. The broker MetricReporter receives the cluster id  after the broker startup is over.
+  * 8. The broker KafkaMetricReporter receives the cluster id  after the broker startup is over.
+  * 9. All the components receive the same cluster id.
+  */
+
+
+object MockConsumerMetricsReporter {
+  val CLUSTER_META: AtomicReference[ClusterResource] = new AtomicReference[ClusterResource]
+}
+
+class MockConsumerMetricsReporter extends MockMetricsReporter with ClusterResourceListener {
+
+  override def onUpdate(clusterMetadata: ClusterResource) {
+    MockConsumerMetricsReporter.CLUSTER_META.set(clusterMetadata)
+  }
+}
+
+object MockProducerMetricsReporter {
+  val CLUSTER_META: AtomicReference[ClusterResource] = new AtomicReference[ClusterResource]
+}
+
+class MockProducerMetricsReporter extends MockMetricsReporter with ClusterResourceListener {
+
+  override def onUpdate(clusterMetadata: ClusterResource) {
+    MockProducerMetricsReporter.CLUSTER_META.set(clusterMetadata)
+  }
+}
+
+object MockBrokerMetricsReporter {
+  val CLUSTER_META: AtomicReference[ClusterResource] = new AtomicReference[ClusterResource]
+}
+
+class MockBrokerMetricsReporter extends MockMetricsReporter with ClusterResourceListener {
```

Comment:
If you have classes that are only used in a particular test, they should live inside a companion object, e.g.

```scala
object EndToEndClusterIdTest {
// classes go here.
}
```

If they are used by multiple tests, they should live in their own files.